### PR TITLE
Support sub-menu for dropdown

### LINKED ISSUE #212: Drop-down menus in navbar: support sub-menu items
v1.7.2

current: drop down menus in the top nav bar supports only one level
e.g.,
image

suggested: support multiple levels

### PR DESCRIPTION
**Pull request by jonahtanjz**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

**Overview of changes:**
Resolves #212.

Sub-menus can be added by nesting dropdowns so that users do not have to remember a new syntax. A new component `<submenu>` is created to handle the logic. `<dropdown>` component is edited such that if a `<dropdown>` is nested, it will be converted to a `<submenu>`. Sub-menus can also contain sub-menus using the same syntax for multiple levels. Currently, sub-menus will appear to the right of a dropdown on desktop version and becomes an accordion on mobile.

**Anything you'd like to highlight / discuss:**
Implementation.

**Testing instructions:**
Sub-menus can be added by nesting `<dropdown>`.

For example:
```
<dropdown>
    <li>...</li> // normal dropdown item
    <li>...</li>
    <dropdown> // will be converted to submenu
        <li>...</li> // submenu item
        <li>...</li>
        <dropdown> // 2nd-level submenu
            <li>...</li> // 2nd-level submenu item
            <li>...</li>
        </dropdown>
    </dropdown>
</dropdown>

```
A dropdown with 2 levels of submenu will be created.

**Proposed commit message: (wrap lines at 72 characters)**
Add sub-menu support for dropdown

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

### CODE REVIEW

**Comment by jonahtanjz** [comment]
@ang-zeyu this is ready for review and do let me know on how this can be improved on. If the implementation is fine, will proceed to add in tests and documentation as well.

**Review by ang-zeyu — COMMENTED** [review]
Just a review on the overall approach / issues:

1. I was thinking of using the https://v3.vuejs.org/guide/component-provide-inject.html#working-with-reactivity. 

e.g.

you could do
```
<dropdown>
  <div class="some-wrapper">
    <dropdown></dropdown> // would still work when nested
  </div>
</dropdown>
```

the approach here is more consistent with vue-strap's, but as you might have observed its a rather low level approach that ties itself heavily to the dom. (note vue-strap didn't have this api at the time)

But I would be ok with accepting this as quite a bit of work has been done already. If you're up for it, let's change this over to provide/inject (could even make a separate PR to shift the original implementation to it first if it makes things easier)

2. Positioning of submenus needs to be dynamic (we can assume right position first then flip it over to the left otherwise). (note the submenu may overflow over the screen currently) You could explore the [popperjs](https://popper.js.org/) library to do this (bootstrap-vue uses this internally for its popovers,dropdowns...)

3. Really small nit: Make the submenu show on-hover as well? 🤔

**Comment by ang-zeyu** [comment]
One more note since you're tackling #980 as well: you may want to finish shifting the mobile navbar design to it first, as dropdowns contained within navbars look / behave quite differently - less work to do here

**Comment by jonahtanjz** [comment]
@ang-zeyu I've made some changes accordingly.
> I was thinking of using the https://v3.vuejs.org/guide/component-provide-inject.html#working-with-reactivity.

I didn't use provide/inject in the end as when I was trying to implement this, I managed to remove all the unnecessary communication between the `dropdown` and `submenu` components. Not sure where else to use it as the rest of the search of the virtual DOM seems necessary, for example listening for clicks on `li>a` as these are user defined.

> Positioning of submenus needs to be dynamic (we can assume right position first then flip it over to the left otherwise). (note the submenu may overflow over the screen currently) You could explore the popperjs library to do this (bootstrap-vue uses this internally for its popovers,dropdowns...)

I tried out popperjs but the auto positioning didn't work very well/a little buggy. I switched to using a modified bootstrap javascript which similarly solves this issue. Hopefully that's fine ;)

I've also added tests and updated the docs.

File: packages/vue-components/src/Submenu.vue [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import $ from './utils/NodeList';
+import positionSubmenu from './utils/submenu';
+
+export default {
+  props: {
+    show: {
+      type: Boolean,
+      default: false,
+    },
+    'class': null,
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:33]: 👀

File: packages/vue-components/src/Submenu.vue [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  },
+  data() {
+    return {
+      dropright: true,
+      dropleft: false,
+    };
+  },
+  computed: {
+    classes() {
+      return [
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:47]: let's place this directly in the template only for consistency with other components'

File: docs/userGuide/syntax/dropdowns.mbdf [hunk 3]
```diff
@@ -56,6 +56,23 @@
 </variable>
 </include>
 
+**You can also use create multi-level submenu by nesting Dropdowns.**
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">html</variable>
+<variable name="code">
+<!-- Can simply nest the dropdown syntax -->
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/dropdowns.mbdf:64]:
```suggestion
<!-- Nest the dropdown syntax to create dropdown submenus -->
```

File: packages/vue-components/src/Navbar.vue [hunk 4]
```diff
@@ -126,9 +126,19 @@ export default {
...
-    addClassIfDropdown(dropdownLinks, a) {
+    addClassIfDropdown(dropdownLinks, a, li) {
       if (dropdownLinks.includes(a)) {
         a.classList.add('dropdown-current');
+        this.addClassIfSubmenu(a, li);
+      }
+    },
+    addClassIfSubmenu(a, li) {
+      let el = a.parentElement;
+      while (el != li) {
```
Comment (ang-zeyu) [inline packages/vue-components/src/Navbar.vue:137]: `!==`

File: packages/vue-components/src/Submenu.vue [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    showBool() {
+      return toBoolean(this.show);
+    },
+    slots() {
+      return this.$slots.default;
+    },
+  },
+  methods: {
+    hideSubmenu() {
+      this.show = false;
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:64]: let's use `data()` for properties that could change

File: packages/vue-components/src/Submenu.vue [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+<style scoped>
+.dropdown-submenu {
+  color: #212529 !important;
+  padding: 0 !important;
+  position: relative;
+}
+
+@media (max-width: 767px) {
+  .dropdown-submenu > ul {
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:132]: hmm can this be moved to the same media query below?

**Review by ang-zeyu — COMMENTED** [review]
> I didn't use provide/inject in the end as when I was trying to implement this, I managed to remove all the unnecessary communication between the `dropdown` and `submenu` components.

Not so much for removing unnecessary communication but just allowing parent / child communication across different [compilation scopes](https://vuejs.org/v2/guide/components-slots.html#Compilation-Scope). 

Nice work isolating the implementation as much as possible - only have one not-so-beneficial suggestion where this can be used now. (see comment below)

You could take a look at the following for examples how it could be used more beneficially:
- [bv components](https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/dropdown/dropdown-item.js)
- our own `<quiz>` / `<question>` / `<q-option>` components

> Not sure where else to use it as the rest of the search of the virtual DOM seems necessary, for example listening for clicks on `li>a` as these are user defined.

yes this is still necessary unfortunately, due to our use of standard dom elements.

> I tried out popperjs but the auto positioning didn't work very well/a little buggy. I switched to using a modified bootstrap javascript which similarly solves this issue. Hopefully that's fine ;)

Seems ok too 👍 

---


Some strange behaviours:

dropdown-sub
 (main / submenu closing on hover)

Untitled
(this is on mobile view)

Some suggestions:

**Comment by jonahtanjz** [comment]
@ang-zeyu Thanks for the suggestions and clarification 😄. I've updated it accordingly.

> Some strange behaviours:
> (main / submenu closing on hover)

For the main menu, it is closing because when the inner children of the submenu toggle is clicked/hovered, it triggers an `blur` event. I've fixed this by preventing the event from bubbling up. 

As for the submenu closing on hover, it is because when the user hovers over the submenu toggle, it will mirror a click. That is the reason why the second hover will close the submenu. One workaround is to only allow submenus to be opened by hovering over and not by clicking. This is the current implementation for now.

File: packages/vue-components/src/Submenu.vue (hunk 6 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue]: Thanks! Updated :)

File: packages/vue-components/src/Submenu.vue (hunk 5 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue:59]: Updated :)

File: packages/vue-components/src/Navbar.vue (hunk 4 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Navbar.vue]: Thanks! Updated :)

File: docs/userGuide/syntax/dropdowns.mbdf (hunk 3 above)
Comment (jonahtanjz) [reply docs/userGuide/syntax/dropdowns.mbdf]: Updated :)

File: packages/vue-components/src/Submenu.vue (hunk 2 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue]: Alright. Updated :)

File: packages/vue-components/src/Submenu.vue (hunk 1 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue]: Thanks! Updated :)

File: packages/vue-components/src/Dropdown.vue [hunk 7]
```diff
@@ -72,6 +79,7 @@ export default {
       return toBoolean(this.disabled);
     },
     isLi () { return this.$parent._navbar || this.$parent.menu || this.$parent._tabset },
+    isSubmenu() { return this.$parent._dropdown || this.$parent._submenu },
```
[the remark this answers is not in the document]
Comment (jonahtanjz) [reply packages/vue-components/src/Dropdown.vue]: Thanks for the clarification. Have added this in :)

File: packages/vue-components/src/Dropdown.vue [hunk 8]
```diff
@@ -13,6 +13,13 @@
       </ul>
     </slot>
   </li>
+  <submenu v-else-if="isSubmenu" ref="submenu">
+    <slot slot="button" name=button></slot>
```
[the remark this answers is not in the document]
Comment (jonahtanjz) [reply packages/vue-components/src/Dropdown.vue]: Thanks! Updated :)

File: packages/vue-components/src/Dropdown.vue [hunk 9]
```diff
@@ -112,7 +120,7 @@ export default {
     showDropdownMenu() {
       this.show = true;
       $(this.$refs.dropdown).findChildren('ul').each(ul => ul.classList.toggle('show', true));
-    },
+    }
```
Comment (ang-zeyu) [inline packages/vue-components/src/Dropdown.vue:123]: let's correct these in another PR; we could enable linting for `dropdown.vue` as well. (the unlinted ones are files that haven't changed much from vue-strap)

File: packages/vue-components/src/Submenu.vue [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+      show: false,
+      dropright: true,
+      dropleft: false,
+    };
+  },
+  computed: {
+    disabledBool() {
+      return toBoolean(this.disabled);
+    },
+    showBool() {
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:53]: this (could replace with just `show`) and `slots()` dosen't seem used

File: packages/vue-components/src/Submenu.vue [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    const $el = $(this.$refs.submenu);
+    $el.offBlur();
+    $el.findChildren('a,button').off();
+    $el.findChildren('ul').off();
+  },
+};
+</script>
+
+<style scoped>
+.dropdown-submenu {
```
Comment (ang-zeyu) [inline packages/vue-components/src/Submenu.vue:126]: let's indent these to fit the stylelint rules too (they've never been validated in our scripts for `.vue` files but should be)

**Review by ang-zeyu — COMMENTED** [review]
Couple more nits:

File: packages/vue-components/src/Dropdown.vue (hunk 9 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Dropdown.vue:123]: Alright sure will open another PR for this 👍

File: packages/vue-components/src/Submenu.vue (hunk 10 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue]: Thanks!. Replaced all `showBool` with `show`

File: packages/vue-components/src/Submenu.vue (hunk 11 above)
Comment (jonahtanjz) [reply packages/vue-components/src/Submenu.vue]: Updated :)

**Comment by jonahtanjz** [comment]
@ang-zeyu PR is updated accordingly :)

**Review by ang-zeyu — APPROVED** [review]
Lgtm 👍

**Pull request merged**